ext/php8/comms_php.c

```diff
@@ -1,57 +1,74 @@
 #include "comms_php.h"
 
+#include <SAPI.h>
 #include <stdatomic.h>
 
 #include "arrays.h"
 #include "compat_string.h"
 #include "compatibility.h"
 #include "coms.h"
+#include "ddshared.h"
 #include "ddtrace.h"
 #include "logging.h"
 #include "mpack/mpack.h"
 
 ZEND_EXTERN_MODULE_GLOBALS(ddtrace);
 
-static void _comms_convert_append(zval *item, size_t offset, void *context) {
-    struct curl_slist **list = context;
-    zval converted;
-    TSRMLS_FETCH();
+static void dd_append_header(struct curl_slist **list, const char *key, const char *val) {
+    /* The longest Agent header should be:
+     * Datadog-Container-Id: <64-char-hash>
+     * So 256 should give us plenty of wiggle room.
+     */
+    char header[256];
+    size_t len = snprintf(header, sizeof(header), "%s: %s", key, val);
+    if (len > 0 && len < sizeof(header)) {
+        *list = curl_slist_append(*list, header);
+    }
+}
 
-    PHP5_UNUSED(offset);
-    PHP7_UNUSED(offset);
+static void dd_append_agent_headers(struct curl_slist **list) {
+    dd_append_header(list, "Datadog-Meta-Lang", "php");
+    dd_append_header(list, "Datadog-Meta-Version", PHP_VERSION);
+    dd_append_header(list, "Datadog-Meta-Lang-Interpreter", sapi_module.name);
+    dd_append_header(list, "Datadog-Meta-Tracer-Version", PHP_DDTRACE_VERSION);
+
+    datadog_string *id = ddshared_container_id();
+    if (id != NULL) {
+        dd_append_header(list, "Datadog-Container-Id", id->val);
+    }
 
-    ddtrace_convert_to_string(&converted, item TSRMLS_CC);
-    *list = curl_slist_append(*list, Z_STRVAL(converted));
-    zval_dtor(&converted);
+    /* Curl will add Expect: 100-continue if it is a POST over a certain size. The trouble is that CURL will
+     * wait for *1 second* for 100 Continue response before sending the rest of the data. This wait is
+     * configurable, but requires a newer curl than we have on CentOS 6. So instead we send an empty Expect.
+     */
+    dd_append_header(list, "Expect", "");
 }
 
-static struct curl_slist *_dd_convert_hashtable_to_curl_slist(HashTable *input) {
-    if (zend_hash_num_elements(input) > 0) {
-        struct curl_slist *list = NULL;
-        ddtrace_array_walk(input, _comms_convert_append, &list);
-        return list;
+static bool dd_memoize_http_headers(void) {
```

Comment:
Now that this is done at the extension level and all this data is static, we don't need to memoize it nor do atomics; we can generate it when the process starts but before the background sender is started. Clean up the headers after the background thread has been cleaned up in process shutdown.

Reply:
Nice catch. This will greatly simplify Agent HTTP header initialization.